File: bindings/javascript/tests/migration.test.ts
import { afterEach, describe, expect, test } from "bun:test";
import { copyFileSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import {
  AsyncDatabase,
  Database,
  migrateDatabase,
  probeFormat,
  Step,
  ZovaError,
} from "../js/index.js";

const temporaryDirectories: string[] = [];

function temporaryDirectory(): string {
  const directory = mkdtempSync(join(tmpdir(), "zova-javascript-migration-"));
  temporaryDirectories.push(directory);
  return directory;
}

function format9FixtureCopy(name: string): string {
  const directory = temporaryDirectory();
  const source = join(directory, name);
  copyFileSync(
    join(import.meta.dir, "..", "..", "..", "tests", "fixtures", "format-9.zova"),
    source,
  );
  return source;
}

function sameBytes(a: Uint8Array, b: Uint8Array): boolean {
  return a.byteLength === b.byteLength && a.every((byte, index) => byte === b[index]);
}

afterEach(() => {
  for (const directory of temporaryDirectories.splice(0)) {
    rmSync(directory, { force: true, recursive: true });
  }
});

describe("Storage-format migration", () => {
  test("probeFormat classifies migratable and current formats", () => {
    const source = format9FixtureCopy("probe-source.zova");
    const before = readFileSync(source);

    const info = probeFormat(source);
    expect(info.formatVersion).toBe(9);
    expect(info.compatibility).toBe("migratable");

    const current = join(temporaryDirectory(), "current.zova");
    const database = Database.create(current);
    database.exec("create table t(id integer)");
    database.close();

    const currentInfo = probeFormat(current);
    expect(currentInfo.formatVersion).toBe(10);
    expect(currentInfo.compatibility).toBe("current");

    expect(sameBytes(readFileSync(source), before)).toBe(true);
  });

  test("probeFormat rejects non-Zova paths", () => {
    const path = join(temporaryDirectory(), "not-zova.txt");
    writeFileSync(path, "not a zova file");
    expect(() => probeFormat(path)).toThrow(ZovaError);
    try {
      probeFormat(path);
    } catch (error) {
      expect((error as ZovaError).code).toBe("ZOVA_NOT_ZOVA_PATH");
    }
  });

  test("migrateDatabase copies the fixture forward and reopens", () => {
    const source = format9FixtureCopy("migrate-source.zova");
    const destination = join(temporaryDirectory(), "migrate-destination.zova");
    const before = readFileSync(source);

    migrateDatabase(source, destination);

    const info = probeFormat(destination);
    expect(info.formatVersion).toBe(10);
    expect(info.compatibility).toBe("current");
    expect(sameBytes(readFileSync(source), before)).toBe(true);

    const database = Database.open(destination);
    const statement = database.prepare("select count(*) from user_documents");
    expect(statement.step()).toBe(Step.Row);
    expect(statement.columnInteger(0)).toBe(3n);
    expect(statement.step()).toBe(Step.Done);
    statement.close();
    database.exec("create table post_migration(id integer primary key, note text)");
    database.close();
  });

  test("migrateDatabase refuses existing destinations", () => {
    const source = format9FixtureCopy("migrate-fail-source.zova");
    const destination = join(temporaryDirectory(), "occupied.zova");
    writeFileSync(destination, "occupied");
    try {
      migrateDatabase(source, destination);
      throw new Error("expected migrateDatabase to fail");
    } catch (error) {
      expect(error).toBeInstanceOf(ZovaError);
      expect((error as ZovaError).code).toBe("ZOVA_DESTINATION_EXISTS");
    }
  });

  test("migrateDatabase reports no migration path for current formats", () => {
    const directory = temporaryDirectory();
    const current = join(directory, "current-source.zova");
    const destination = join(directory, "current-destination.zova");
    const database = Database.create(current);
    database.exec("create table t(id integer)");
    database.close();

    try {
      migrateDatabase(current, destination);
      throw new Error("expected migrateDatabase to fail");
    } catch (error) {
      expect(error).toBeInstanceOf(ZovaError);
      expect((error as ZovaError).code).toBe("ZOVA_NO_MIGRATION_PATH");
    }
    expect(() => readFileSync(destination)).toThrow();
  });

  test("async probe and migration run off the event loop and match sync results", async () => {
    const source = format9FixtureCopy("async-migrate-source.zova");
    const destination = join(temporaryDirectory(), "async-migrate-destination.zova");

    const info = await AsyncDatabase.probeFormat(source);
    expect(info.formatVersion).toBe(9);
    expect(info.compatibility).toBe("migratable");

    await AsyncDatabase.migrateDatabase(source, destination);

    const migratedInfo = await AsyncDatabase.probeFormat(destination);
    expect(migratedInfo.formatVersion).toBe(10);
    expect(migratedInfo.compatibility).toBe("current");

    const syncInfo = probeFormat(destination);
    expect(syncInfo.formatVersion).toBe(migratedInfo.formatVersion);
    expect(syncInfo.compatibility).toBe(migratedInfo.compatibility);
  });
});

File: bindings/go/migration_test.go
package zova

import (
	"os"
	"path/filepath"
	"testing"
)

func migrationFixturePath(t *testing.T, name string) string {
	t.Helper()
	return filepath.Join("..", "..", "tests", "fixtures", name)
}

func copyMigrationFixture(t *testing.T, name string) string {
	t.Helper()
	source := migrationFixturePath(t, name)
	destination := tempZovaPath(t, "migrate-"+name)
	bytes, err := os.ReadFile(source)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	if err := os.WriteFile(destination, bytes, 0o644); err != nil {
		t.Fatalf("write fixture copy: %v", err)
	}
	t.Cleanup(func() { _ = os.Remove(destination) })
	return destination
}

func readAllBytes(t *testing.T, path string) []byte {
	t.Helper()
	bytes, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	return bytes
}

func sameBytes(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for index := range a {
		if a[index] != b[index] {
			return false
		}
	}
	return true
}

func TestProbeFormatClassifiesFormats(t *testing.T) {
	fixture := migrationFixturePath(t, "format-9.zova")
	before := readAllBytes(t, fixture)

	info, err := ProbeFormat(fixture)
	if err != nil {
		t.Fatalf("ProbeFormat(format-9) = %v", err)
	}
	if info.FormatVersion != 9 || info.Compatibility != FormatMigratable {
		t.Fatalf("ProbeFormat(format-9) = %+v, want format 9 migratable", info)
	}
	if name := FormatCompatibilityName(info.Compatibility); name != "migratable" {
		t.Fatalf("compatibility name = %q, want migratable", name)
	}

	current := tempZovaPath(t, "probe-current")
	db, err := Create(current)
	if err != nil {
		t.Fatalf("Create = %v", err)
	}
	if err := db.Close(); err != nil {
		t.Fatalf("Close = %v", err)
	}
	info, err = ProbeFormat(current)
	if err != nil {
		t.Fatalf("ProbeFormat(current) = %v", err)
	}
	if info.FormatVersion != 10 || info.Compatibility != FormatCurrent {
		t.Fatalf("ProbeFormat(current) = %+v, want format 10 current", info)
	}
	if name := FormatCompatibilityName(info.Compatibility); name != "current" {
		t.Fatalf("compatibility name = %q, want current", name)
	}

	if after := readAllBytes(t, fixture); !sameBytes(before, after) {
		t.Fatalf("ProbeFormat mutated the fixture")
	}
}

func TestProbeFormatRejectsNonZovaPaths(t *testing.T) {
	path := filepath.Join(t.TempDir(), "not-zova.txt")
	if err := os.WriteFile(path, []byte("not a zova file"), 0o644); err != nil {
		t.Fatalf("write: %v", err)
	}
	if _, err := ProbeFormat(path); !errorStatusIs(err, StatusNotZovaPath) {
		t.Fatalf("ProbeFormat(not-zova) = %v, want StatusNotZovaPath", err)
	}
}

func TestMigrateFixtureForwardAndReopen(t *testing.T) {
	source := copyMigrationFixture(t, "format-9.zova")
	destination := tempZovaPath(t, "migrate-destination")
	before := readAllBytes(t, source)

	if err := MigrateDatabase(source, destination); err != nil {
		t.Fatalf("MigrateDatabase = %v", err)
	}

	info, err := ProbeFormat(destination)
	if err != nil {
		t.Fatalf("ProbeFormat(destination) = %v", err)
	}
	if info.FormatVersion != 10 || info.Compatibility != FormatCurrent {
		t.Fatalf("ProbeFormat(destination) = %+v, want format 10 current", info)
	}
	if after := readAllBytes(t, source); !sameBytes(before, after) {
		t.Fatalf("MigrateDatabase mutated the source")
	}

	db, err := Open(destination)
	if err != nil {
		t.Fatalf("Open(destination) = %v", err)
	}
	defer func() { _ = db.Close() }()

	stmt, err := db.Prepare("select count(*) from user_documents")
	if err != nil {
		t.Fatalf("Prepare = %v", err)
	}
	defer func() { _ = stmt.Close() }()
	if step, err := stmt.Step(); err != nil || step != StepRow {
		t.Fatalf("Step = %v, %v; want StepRow", step, err)
	}
	if count, err := stmt.ColumnInt64(0); err != nil || count != 3 {
		t.Fatalf("user_documents count = %d, %v; want 3", count, err)
	}

	if err := db.Exec("create table post_migration(id integer primary key, note text)"); err != nil {
		t.Fatalf("Exec = %v", err)
	}
}

func TestMigrateNoVerifyFlag(t *testing.T) {
	source := copyMigrationFixture(t, "format-9.zova")
	destination := tempZovaPath(t, "migrate-no-verify")

	if err := MigrateDatabase(source, destination, MigrateOptions{NoVerify: true}); err != nil {
		t.Fatalf("MigrateDatabase(no-verify) = %v", err)
	}
	info, err := ProbeFormat(destination)
	if err != nil {
		t.Fatalf("ProbeFormat = %v", err)
	}
	if info.FormatVersion != 10 || info.Compatibility != FormatCurrent {
		t.Fatalf("ProbeFormat = %+v, want format 10 current", info)
	}
}

func TestMigrateFailureStatuses(t *testing.T) {
	source := copyMigrationFixture(t, "format-9.zova")
	destination := tempZovaPath(t, "migrate-fail-destination")
	if err := os.WriteFile(destination, []byte("occupied"), 0o644); err != nil {
		t.Fatalf("write: %v", err)
	}

	if err := MigrateDatabase(source, destination); !errorStatusIs(err, StatusDestinationExists) {
		t.Fatalf("MigrateDatabase(existing) = %v, want StatusDestinationExists", err)
	}
	if err := MigrateDatabase(tempZovaPath(t, "missing"), destination); !errorStatusIs(err, StatusCantOpen) {
		t.Fatalf("MigrateDatabase(missing) = %v, want StatusCantOpen", err)
	}
}
